### PR TITLE
feat(input, textarea): ionInput and ionChange pass event and value

### PR DESCRIPTION
**Pull request by liamdebeasi**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue URL: N/A Internal Ticket

As a result of our recent `ionChange` revisions, the team noted that the `ionInput` and `ionChange` payloads were inconsistent. `ionChange` event detail has an object containing the value, but the `ionInput` event detail has the native event. We were concerned that those switching from `ionChange` to `ionInput` may find this migration confusing due to the different information available across events.

## What is the new behavior?

- `ionInput` and `ionChange` for `ion-input` and `ion-textarea` now have a `detail` payload of an object that contains the current component `value` and the native event (if one exists).
- Created separate `ChangeEventDetail` and `InputEventDetail` interfaces. I opted to keep them separate for now despite them having the same contents. This was done to make the interfaces easier to reason above. We can consolidate them in the future if needed.
- Updated `InputCustomEvent` and `TextareaCustomEvent` to use a generic for `detail` so that developers can specify either the input or the change event detail payload. I made it default to the `{Input,Textarea}ChangeEventDetail` interface to avoid another breaking change, but I am open to discussing this. This change is not needed at the moment, but it will be needed in case the `InputEventDetail` or `ChangeEventDetail` interfaces need to diverge again in the future. 

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

**Review by amandaejohnston — APPROVED** [review]

File: BREAKING.md [hunk 1]
```diff
@@ -92,6 +92,8 @@ This section details the desktop browser, JavaScript framework, and mobile platf
 
 - The `debounce` property has been updated to control the timing in milliseconds to delay the event emission of the `ionInput` event after each keystroke. Previously it would delay the event emission of `ionChange`.
 
+- The `detail` payload for the `ionInput` event now contains an object containing the current `value` as well as the native event that triggered `ionInput`.
```
Comment (sean-perkins) [inline BREAKING.md:95]:
```suggestion
- The `detail` payload for the `ionInput` event now contains an object with the current `value` as well as the native event that triggered `ionInput`.
```

File: BREAKING.md [hunk 2]
```diff
@@ -164,7 +166,7 @@ Developers using these components will need to migrate to using Swiper.js direct
 
 - The `debounce` property has been updated to control the timing in milliseconds to delay the event emission of the `ionInput` event after each keystroke. Previously it would delay the event emission of `ionChange`.
 
-- `ionInput` dispatches an event detail of `null` when the textarea is cleared as a result of `clear-on-edit="true"`.
+- The `detail` payload for the `ionInput` event now contains an object containing the current `value` as well as the native event that triggered `ionInput`.
```
Comment (sean-perkins) [inline BREAKING.md:169]:
```suggestion
- The `detail` payload for the `ionInput` event now contains an object with the current `value` as well as the native event that triggered `ionInput`.
```

**Review by sean-perkins — APPROVED** [review]
Looks good to me!

One small suggestion on the breaking change wording.

**Comment by piotr-cz** [comment]
I have an question - why `InputChangeEventDetail` interface have been changed, so that value may be a `number`:

https://github.com/ionic-team/ionic-framework/commit/eea6ba996ce0f71546f5a14109d0d279400a27e5#diff-5172cfa7e4dd36c59ff7084a0fcea91a7b910956c334673bb22eb138f84a8bda

https://github.com/ionic-team/ionic-framework/blob/eea6ba996ce0f71546f5a14109d0d279400a27e5/core/src/components/input/input-interface.ts#L2

Value of HTMLInputElement may only be a `string` - see [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement)

This change hasn't been described in this PR or Ionic 7 migration docs, but in effect when using TypeScript in event handlers there must be added extra type check

**Comment by liamdebeasi** [comment]
This looks like a typo. I created a PR to fix this here: https://github.com/ionic-team/ionic-framework/pull/27226

**Pull request merged**